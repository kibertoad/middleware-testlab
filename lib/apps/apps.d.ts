export interface EndpointDefinition {
  method: 'get' | 'post' | 'delete' | 'put' | 'patch' | 'options' | 'head'
  path: string
}
