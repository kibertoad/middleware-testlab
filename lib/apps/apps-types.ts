import { Request, Response } from 'express-serve-static-core'
import { BaseContext } from 'koa'

export interface EndpointDefinition {
  method: 'get' | 'post' | 'delete' | 'put' | 'patch' | 'options' | 'head'
  path: string
}

export interface ExpressEndpointAssertor {
  (request: Request): void
}

export interface ExpressEndpointResponseAssertor {
  (response: Response): void
}

export interface ErrorAssertor {
  (error: any): void
}

export interface KoaEndpointAssertor {
  (ctx: BaseContext): void
}
