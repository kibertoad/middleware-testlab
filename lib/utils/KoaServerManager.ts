import { Server } from 'net'
// @ts-ignore
import Application from 'koa'

export class KoaServerManager {
  private server: Server | undefined

  public startBeforeEach(app: Application, port: number = 8888): void {
    beforeEach(() => {
      this.stop()
      this.server = app.listen(port)
    })
  }

  public startBeforeAll(app: Application, port: number = 8888): void {
    beforeEach(() => {
      this.stop()
      this.server = app.listen(port)
    })
  }

  public stopAfterEach(): void {
    afterEach(() => {
      this.stop()
    })
  }

  public stopAfterAll(): void {
    afterAll(() => {
      this.stop()
    })
  }

  private stop(): void {
    if (this.server) {
      this.server.close()
      this.server = undefined
    }
  }
}
