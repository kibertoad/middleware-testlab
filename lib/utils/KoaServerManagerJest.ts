import { Server } from 'net'
// @ts-ignore
import Application from 'koa'
import supertest, { SuperTest, Test } from 'supertest'

export class KoaServerManagerJest {
  private server: Server | undefined
  private readonly port: number

  public constructor(port?: number) {
    this.port = port || 8888
  }

  public request(): SuperTest<Test> {
    return supertest(this.server)
  }

  public start(app: Application, port?: number): void {
    this.stop()
    this.server = app.listen(port || this.port)
  }

  public startBeforeEach(app: Application, port?: number): void {
    beforeEach(() => {
      this.start(app, port)
    })
  }

  public startBeforeAll(app: Application, port?: number): void {
    beforeAll(() => {
      this.start(app, port)
    })
  }

  public stopAfterEach(): void {
    afterEach(() => {
      this.stop()
    })
  }

  public stopAfterAll(): void {
    afterAll(() => {
      this.stop()
    })
  }

  private stop(): void {
    if (this.server) {
      this.server.close()
      this.server = undefined
    }
  }
}
