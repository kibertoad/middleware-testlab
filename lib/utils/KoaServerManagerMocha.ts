import { Server } from 'net'
// @ts-ignore
import Application from 'koa'
import supertest, { SuperTest, Test } from 'supertest'

export class KoaServerManagerMocha {
  private server: Server | undefined

  public request(): SuperTest<Test> {
    return supertest(this.server)
  }

  public start(app: Application, port: number = 8888): void {
    this.stop()
    this.server = app.listen(port)
  }

  public startBeforeEach(app: Application, port: number = 8888): void {
    beforeEach(() => {
      this.start(app, port)
    })
  }

  public startBefore(app: Application, port: number = 8888): void {
    // @ts-ignore
    before(() => {
      this.start(app, port)
    })
  }

  public stopAfterEach(): void {
    afterEach(() => {
      this.stop()
    })
  }

  // afterAll for mocha
  public stopAfter(): void {
    // @ts-ignore
    after(() => {
      this.stop()
    })
  }

  private stop(): void {
    if (this.server) {
      this.server.close()
      this.server = undefined
    }
  }
}
