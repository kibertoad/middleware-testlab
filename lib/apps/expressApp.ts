import { Application, Request, Response } from 'express'
import express, { Router } from 'express'
import { RequestHandlerParams } from 'express-serve-static-core'
import { EndpointDefinition } from './apps'

export const DEFAULT_ENDPOINT = '/'
const DEFAULT_HANDLER = (_req: Request, res: Response, next: Function) => {
  try {
    res.status(201).send
  } catch (e) {
    console.error('Error while processing request: ', e)
    next(e)
  }
}

/**
 *
 * @param {RequestHandlerParams[]} middleware
 * @param {RequestHandlerParams} [endpointHandler] - if handler is not specified, by default sends 204 NO CONTENT
 * @param {string | EndpointDefinition} [endpoint='/'] - if endpoint is passed as string, or not specified, by default GET method is used.
 */
export function newExpressApp(
  middleware: RequestHandlerParams[],
  endpointHandler: RequestHandlerParams = DEFAULT_HANDLER,
  endpoint: string | EndpointDefinition = DEFAULT_ENDPOINT
): Application {
  const app = express()
  middleware.forEach((middlewareEntry: RequestHandlerParams) => {
    app.use(middlewareEntry)
  })

  // Resolve endpoint from params
  let method: string
  let path: string
  if (typeof endpoint === 'string') {
    method = 'get'
    path = endpoint
  } else {
    method = endpoint.method
    path = endpoint.path
  }

  const router = Router()
  // @ts-ignore
  router[method](path, endpointHandler)
  app.use(router)
  return app
}
