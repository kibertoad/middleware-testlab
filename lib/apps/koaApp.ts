import Koa, { BaseContext, Middleware } from 'koa'
import Router from 'koa-router'
import Application from 'koa'
import { EndpointDefinition } from './apps'

export const DEFAULT_ENDPOINT = '/'
const DEFAULT_HANDLER = (ctx: BaseContext, next: Function) => {
  ctx.status = 204
  next()
}

/**
 *
 * @param {Middleware[]} [appMiddleware=[]]
 * @param {Middleware[]} [routerMiddleware=[]]
 * @param {Middleware} [handler] - if handler is not specified, by default sends 204 NO CONTENT
 * @param {string | EndpointDefinition} [endpoint='/'] - if endpoint is passed as string, or not specified, by default GET method is used.
 */
export function newKoaApp({
  appMiddleware = [],
  routerMiddleware = [],
  handler = DEFAULT_HANDLER,
  endpoint = DEFAULT_ENDPOINT
}: {
  appMiddleware?: Middleware[]
  routerMiddleware?: Middleware[]
  handler?: Middleware
  endpoint?: string | EndpointDefinition
}): Application {
  const app = new Koa()
  appMiddleware.forEach((middlewareEntry: Middleware) => {
    app.use(middlewareEntry)
  })

  const router = new Router()

  routerMiddleware.forEach((middlewareEntry: Middleware) => {
    router.use(middlewareEntry)
  })

  // Resolve endpoint from params
  let method: string
  let path: string
  if (typeof endpoint === 'string') {
    method = 'get'
    path = endpoint
  } else {
    method = endpoint.method
    path = endpoint.path
  }

  // @ts-ignore
  router[method](path, handler)
  app.use(router.routes()).use(router.allowedMethods())

  return app
}
