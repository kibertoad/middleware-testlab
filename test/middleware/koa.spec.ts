import { BaseContext } from 'koa'

import { newKoaApp, DEFAULT_ENDPOINT } from '../../lib/apps/koaApp'
import { koaMiddleware } from './koaRequestMutationMiddleware'
import { KoaServerManagerJest } from '../../lib/utils/KoaServerManagerJest'
const port = 8888

describe('mutation routeMiddleware', () => {
  describe('koa', () => {
    const server = new KoaServerManagerJest(port)
    server.stopAfterEach()

    it('happy path', async () => {
      const app = newKoaApp({
        routerMiddleware: [koaMiddleware()],
        handler: (ctx: BaseContext, next: Function) => {
          expect(ctx.logger).toMatchSnapshot()
          ctx.status = 204
          next()
        }
      })
      server.start(app)

      const response = await server.request().get(DEFAULT_ENDPOINT)

      expect(response.status).toEqual(204)
    })
  })
})
