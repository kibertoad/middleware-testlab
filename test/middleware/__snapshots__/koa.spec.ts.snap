// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation routeMiddleware koa happy path 1`] = `
Object {
  "error": [Function],
  "info": [Function],
}
`;
